### PR TITLE
fixed list function implementation

### PR DESCRIPTION
**Pull request by shoaib-moeen**
Fix list() issue mentioned in https://github.com/pyccel/pyccel/issues/1892 . This PR fixes the list() implementation when using a list or a tuple as a parameter

Commit Summary
- Created PythonListFunction class which acts as a wrapper for PythonList and handles such cases.
- Created unit tests using the newly implemented class.

### CODE REVIEW

**Comment by shoaib-moeen** [comment]
/bot run docs spelling pylint pyccel_lint coverage

**Comment by EmilyBourne** [comment]
/bot run docs spelling pylint pyccel_lint coverage

**Comment by shoaib-moeen** [comment]
/bot run docs spelling pylint pyccel_lint coverage

**Comment by shoaib-moeen** [comment]
/bot run docs spelling pylint pyccel_lint coverage

**Comment by shoaib-moeen** [comment]
@EmilyBourne , I could not find the authors file. Can you tell me where I have to update it? Everything else on the checklist is completed

**Comment by Farouk-Echaref** [comment]
@shoaib-moeen it's in the root directory called AUTHORS

**Comment by shoaib-moeen** [comment]
/bot run docs spelling pylint pyccel_lint coverage

File: tests/epyccel/test_lists.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+def test_lists(test_func, language):
+    f1 = test_func
+    f2 = epyccel( f1 , language=language)
+
+    python_out = f1()
+    pyccel_out = f2()
+    print(pyccel_out)
+    print(python_out)
+
+    assert python_out == pyccel_out
```
Comment (EmilyBourne) [inline tests/epyccel/test_lists.py:22]: We already have a file for testing list functions. It is here:
https://github.com/pyccel/pyccel/blob/devel/tests/epyccel/test_epyccel_lists.py

Please can you move this function there so we have all list functions in one place

File: pyccel/ast/builtins.py [hunk 2]
```diff
@@ -715,6 +716,32 @@ def is_homogeneous(self):
...
+    __slots__ = ()
+    _attribute_nodes = ()
+
+    def __new__(cls, arg):
+        if isinstance(arg, PythonList):
+            return arg
+        elif isinstance(arg.shape[0], LiteralInteger):
+            return PythonList(*[arg[i] for i in range(arg.shape[0])])
+        else:
+            raise TypeError(f"Can't unpack {arg} into a list")
```
Comment (EmilyBourne) [inline pyccel/ast/builtins.py:742]: As implemented this function will only handle arguments of known size. For example the following code won't translate:
```python
def f(a : list[int]):
    return list(a)
```
it raises:
```
ERROR at annotation (semantic) stage
pyccel:
 |fatal [semantic]: tmp_list_copy.py [1,10]| Cannot unpack a into a list
Pyccel has encountered syntax that has not been implemented yet. Please create an issue at https://github.com/pyccel/pyccel/issues and provide a small example of your problem. Do not forget to specify your target language. (list[int])
```

The fix should be similar to what is done in #1897 .

File: CHANGELOG.md [hunk 3]
```diff
@@ -36,6 +36,7 @@ All notable changes to this project will be documented in this file.
 -   #1853 : Fix translation of a file whose name conflicts with Fortran keywords.
 -   Link and mention `devel` branch, not `master`.
 -   #1047 : Print the value of an unrecognised constant.
+-   #1892 : Fix implementation of list function when list or tuple is passed as parameter.
```
Comment (EmilyBourne) [inline CHANGELOG.md:39]: I think this solution is more general than you suggest here. E.g. it should also work for sets. Maybe just mention iterables?
```suggestion
-   #1892 : Fix implementation of list function when an iterable is passed as parameter.
```

File: pyccel/ast/builtins.py (hunk 2 above)
Comment (shoaib-moeen) [reply pyccel/ast/builtins.py:742]: @EmilyBourne , I am trying to fix the issue. After some changes I am getting another exception which is as follows:

 |fatal [semantic]: test.py [2,10]| Cannot handle non-homogenous type index
Pyccel has encountered syntax that has not been implemented yet. Please create an issue at https://github.com/pyccel/pyccel/issues and provide a small example of your problem. Do not forget to specify your target language. (list[int])

As I understand the lists should be homogeneous. Should this PR handle this also?

Comment (EmilyBourne) [reply pyccel/ast/builtins.py:742]: I've had a quick look and it seems the error is coming from [here](https://github.com/pyccel/pyccel/blob/5a75f34563bd1ea3287afc29949bca98400be6fe/pyccel/parser/semantic.py#L1940). This function is designed to handle type annotations. Type annotations for lists are only added in #1847.

Sorry, I should have given you a better test case. Here is a new one:
```python
def f(a : 'int[:]'):
    return list(a)
```
where `a` is a NumPy array. The type annotations are properly implemented here

Comment (EmilyBourne) [reply pyccel/ast/builtins.py:742]: @shoaib-moeen if you merge the main branch the error you were seeing should now be fixed.

**Comment by EmilyBourne** [comment]
@shoaib-moeen Do you plan to come back and finish this PR or can I pick it up and complete it?

**Comment by shoaib-moeen** [comment]
Sorry. I have been busy since last two months. You can go ahead and
complete it.

File: pyccel/ast/builtins.py [hunk 4]
```diff
@@ -719,6 +720,47 @@ def is_homogeneous(self):
...
+    @property
+    def copied_obj(self):
+        """
+        The object being copied.
+
+        The object being copied to create a new dict instance.
+        """
+        return self._args[0]
+
+
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/builtins.py:763]: /bot accept
This will be useful when we add C/Fortran support

File: pyccel/ast/builtins.py [hunk 5]
```diff
@@ -719,6 +720,47 @@ def is_homogeneous(self):
...
+        self._class_type = copied_obj.class_type
+        self._shape = copied_obj.shape
+        super().__init__(copied_obj)
+
+    @property
+    def copied_obj(self):
+        """
+        The object being copied.
+
+        The object being copied to create a new dict instance.
```
Comment (jalalium) [inline pyccel/ast/builtins.py:759]: Shouldn't it be `... to create a new LIST instance` ?

**Review by jalalium — CHANGES_REQUESTED** [review]

File: pyccel/ast/builtins.py (hunk 5 above)
Comment (EmilyBourne) [reply pyccel/ast/builtins.py]: Ooops, bad copy paste. Fixed : 700ab2a0

**Review by jalalium — APPROVED** [review]

**Review by EmilyBourne — APPROVED** [review]

**Pull request merged**